### PR TITLE
Job timeout handling

### PR DESCRIPTION
**Pull request by nulian**

### CODE REVIEW

File: lib/roger/job.ex [hunk 1]
```diff
@@ -133,11 +119,24 @@ defmodule Roger.Job do
   def create(module, args \\ [], id \\ generate_job_id()) when is_atom(module) do
     keys =
       ~w(queue_key execution_key)a
-      |> Enum.map(fn(prop) -> {prop, Kernel.apply(module, prop, [args])} end)
+      |> Enum.map(fn(prop) ->
+        if function_exported?(module, prop, 1) do
```
Comment (Eptis) [inline lib/roger/job.ex:123]: Is there no way around all the `function_exported?/3` checks?

File: lib/roger/partition/worker.ex [hunk 2]
```diff
@@ -48,62 +48,113 @@ defmodule Roger.Partition.Worker do
       channel: channel,
       meta: meta,
       raw_payload: payload}
+    Process.flag(:trap_exit, true)
     {:ok, state, 0}
   end
 
+  def terminate(_reason, state) do
```
Comment (Eptis) [inline lib/roger/partition/worker.ex]: Maybe add some `@spec`s and `@doc`s to the public functions

File: lib/roger/partition/worker.ex [hunk 3]
```diff
@@ -48,62 +48,113 @@ defmodule Roger.Partition.Worker do
...
+    Process.exit(pid, :kill)
+    handle_error(job, {:timeout, "Job stopped because of timeout"}, nil, state)
+    {:stop, :normal, state}
+  end
+
   def handle_info(:timeout, state) do
     case Job.decode(state.raw_payload) do
       {:ok, job} ->
         job = %Job{job | started_at: Roger.now}
         if Global.cancelled?(state.partition_id, job.id, :remove) do
```
Comment (Eptis) [inline lib/roger/partition/worker.ex]: Can this be broken down into smaller functions? It's quite large and somewhat hard to follow

File: lib/roger/partition/worker.ex [hunk 4]
```diff
@@ -48,62 +48,113 @@ defmodule Roger.Partition.Worker do
...
-              GProc.unregp(name(job.id))
-              GProc.unregp({:roger_job_worker_meta, state.partition_id, job.id})
-
+            parent = self()
+            {pid, _ref} = spawn_monitor(fn () ->
+              execute_job(job, state, parent)
+              send(parent, :job_finished)
+            end)
+
+            if is_number(job.max_execution_time) do
```
Comment (Eptis) [inline lib/roger/partition/worker.ex]: Maybe change to `is not :infinity` to make the intent more explicit

File: lib/roger/partition/worker.ex [hunk 5]
```diff
@@ -48,62 +48,113 @@ defmodule Roger.Partition.Worker do
...
     end
+
+  end
+
+  def handle_info(:job_finished, state) do
+    {:stop, :normal, state}
+  end
+
+  def handle_info(:job_errored, state) do
+    GProc.unregp(name(state.job.id))
```
Comment (Eptis) [inline lib/roger/partition/worker.ex]: Pipe it™

File: lib/roger/partition/worker.ex [hunk 6]
```diff
@@ -48,62 +48,113 @@ defmodule Roger.Partition.Worker do
...
+    GProc.unregp({:roger_job_worker_meta, state.partition_id, state.job.id})
```
Comment (Eptis) [inline lib/roger/partition/worker.ex]: What does `GProc.unregp` do exactly?

File: lib/roger/partition/worker.ex [hunk 7]
```diff
@@ -48,62 +48,113 @@ defmodule Roger.Partition.Worker do
...
+    before_run_state = callback(:before_run, [state.partition_id, job])
+    try do
+      result = Job.execute(job)
+
+      job_done(job, :ack, state)
+
+      callback(:after_run, [state.partition_id, job, result, before_run_state])
+    catch
+      type, exception ->
+        #Logger.error "Execution error: #{t}:#{inspect e}"
```
Comment (Eptis) [inline lib/roger/partition/worker.ex]: Should logging remain disabled?

File: lib/roger/partition/worker.ex [hunk 8]
```diff
@@ -48,62 +48,113 @@ defmodule Roger.Partition.Worker do
...
+        handle_error(job, {type, exception}, before_run_state, state)
+        send(parent, :job_errored)
+    end
+  end
+
+  defp handle_error(job, {type, exception}, before_run_state, state) do
+    cb = if Job.retryable?(job) do
+      case Retry.retry(state.channel, state.partition_id, job) do
```
Comment (Eptis) [inline lib/roger/partition/worker.ex]: You can probably combine the if in the case statement as well. The last function executed in a case statement is returned so you either get back a retry response, or `false`

Comment (Eptis) [reply lib/roger/partition/worker.ex]: Or use a `with` expression

File: lib/roger/job.ex (hunk 1 above)
Comment (nulian) [reply lib/roger/job.ex:123]: It's a pity they didn't implement a cleaner way for that I looked up how elixir developers handled it in genserver optional callbacks.

```
 Usually, behaviour modules know if they should call those callbacks based on configuration, or they check if the callbacks are defined with `function_exported?/3` or `macro_exported?/3`.
```

Comment (Eptis) [reply lib/roger/job.ex:123]: And using something like `Code.ensure_loaded` when creating a job?

File: lib/roger/partition/worker.ex (hunk 6 above)
Comment (nulian) [reply lib/roger/partition/worker.ex]: Eerlijk zeggen niet 100% duidelijk maar ik denk deze worker genserver pid uit de gproc registry halen. Ongeveer net als elixir registry zodat je kan ophalen welke workers er bezig zijn.

File: lib/roger/partition/worker.ex (hunk 7 above)
Comment (nulian) [reply lib/roger/partition/worker.ex]: I will remove it you can implement your own logging in the on_error callback.

File: lib/roger/job.ex (hunk 1 above)
Comment (nulian) [reply lib/roger/job.ex:123]: Yeah that might be good check to add. So apps that don't fully embed everything always work.

**Comment by Eptis** [comment]
👍
